tests/func/test_data_cloud.py

```diff
@@ -894,3 +894,79 @@ def test(self):
         assert self.message_header in self._caplog.text
         assert self.message_foo_part in self._caplog.text
         assert self.message_bar_part in self._caplog.text
+
+
+# New pytest style tests parametrized properly
+
+import pytest  # noqa
+from dvc import api  # noqa
+
+
+# NOTE: staticmethod is only needed in Python 2
+class S3:
+    should_test = staticmethod(_should_test_aws)
+    get_url = staticmethod(get_aws_url)
+
+
+class GCP:
+    should_test = staticmethod(_should_test_gcp)
+    get_url = staticmethod(get_gcp_url)
+
+
+class Azure:
+    should_test = staticmethod(_should_test_azure)
+    get_url = staticmethod(get_azure_url)
+
+
+class OSS:
+    should_test = staticmethod(_should_test_oss)
+    get_url = staticmethod(get_oss_url)
+
+
+class SSH:
+    should_test = staticmethod(_should_test_ssh)
+    get_url = staticmethod(get_ssh_url)
+
+
+class HDFS:
+    should_test = staticmethod(_should_test_hdfs)
+    get_url = staticmethod(get_hdfs_url)
+
+
+remote_params = [S3, GCP, Azure, OSS, SSH, HDFS]
+
+
+@pytest.fixture
+def remote(request):
+    if not request.param.should_test():
+        raise pytest.skip()
+    return request.param
+
+
+def pytest_generate_tests(metafunc):
+    if "remote" in metafunc.fixturenames:
+        metafunc.parametrize("remote", remote_params, indirect=True)
+
+
+def test_api_get_url(repo_dir, dvc_repo, remote):
+    remote_url = remote.get_url()
+
+    main(["remote", "add", "-d", "upstream", remote_url])
+    dvc_repo.add(repo_dir.FOO)
+
+    assert api.get_url(repo_dir.FOO) == "%s/%s" % (
+        remote_url,
+        "ac/bd18db4cc2f85cedef654fccc4a4d8",
+    )
+
+
+def test_api_open(repo_dir, dvc_repo, remote):
+    main(["remote", "add", "-d", "upstream", remote.get_url()])
```

Comment:
Main returns an integer, which you should check. Otherwise, you might be missing an error.

Reply:
I have config.remote_add coming in my pkgs patch, which should help us avoid using CLI for config modifications. So it would be something like `dvc_repo.config.remote_add`, so exceptions won't be caught like CLI does.

Reply:
As I said above I can't see complicating config with outer knowledge as a good thing. If I would add a way to modify config I would make it in a way that it won't know about its contents. So:

```python
remote_name = "..."
remote_section = Config.REMOTE_SECTION_FMT.format(remote_name)
config[remote_section] = {...}
config["core"]["remote"] = remote_name
```

somewhere in remote or repo - the same place as we do `.get_remote()`.